### DISCUSSION TITLE
Plugin not working after Build after my development mode

### DISCUSSION THREAD

**Discussion by monukaushik55** [fbaligand/kibana-enhanced-table#304] [General]
Dear Sir,

I have installed plugin in my development mode and I have build using compile-build command without any changes after that build successfully and installed successfully on my production mode. But after that we are click on plugin and we get error "et not used before intilization". screenshot also attached.

Please suggest in this regards.

![2590dce1-1ac6-45b5-a3c3-84b0ba827675](https://user-images.githubusercontent.com/97586649/210502403-8bac8c7b-a865-4a7e-b9f2-4f8fd0d3950f.jpeg)

**Comment by fbaligand** [comment]
Can you detail these elements please:
- which kibana version?
- which enhanced-table branch?
- what exact commands did you run (especially yarn commands)?
- finally, why do you build plugin for yourself instead of using releases provided on github?

**Reply by monukaushik55** [reply]
Thank you sir for your quick response:

**which kibana version? -** 
Production Version : 8.1.3
Development Version : 8.1.0

**which enhanced-table branch?**
By Default Branch : Master

**what exact commands did you run (especially yarn commands)?**
Development Mode:
cd plugins
git clone https://github.com/fbaligand/kibana-enhanced-table.git enhanced-table
cd plugins/kibana-enhanced-table
yarn install

For build: yarn compile-and-build --kibana-version 8.1.3
after build install in production using: bin/kibana-plugin install <path>
(note : build and install successfully )

**finally, why do you build plugin for yourself instead of using releases provided on github?**

Sir, I have developed my custom plugin using visTypeMarkdown plugin in kibana. My plugin properly working in development mode and successfully build (using yarn build command)  but getting error in production after successfully installed. After that I have found only your plugin working perfectly and latest in kibana from all external plugin and I have installed your plugin in my development mode for checking process of build but I have also getting error in enhance table plugin after build from my development mode. Please find attached my plugin error in production mode. I have not understand why getting error in my plugin in production if installed successfully, build successfully and working properly in development mode.

![98babf9c-91c9-45e8-aa11-550aa37a8209](https://user-images.githubusercontent.com/97586649/210765396-bbf2e945-830c-4905-8ecc-341015588660.jpeg)

**Comment by fbaligand** [comment]
Hi,

Thanks for your detailed answer.

So:
- To make it work with Kibana 8.1, you have to use "8.0" branch. And yes, I agree that it is not really documented.
- Concerning the commands that you use, it looks like fine.

**Reply by monukaushik55** [reply]
Sir,

Please reply below mentioned two questions:

1. I have realised enhanced plugin used "**https://jimmywarting.github.io/StreamSaver.js/mitm.html?version=2.0.0**" link for download data in CSV after check on full csv export.
----- In this regards, Please confirm this is safe/secure or not. If this is not secure please suggest another solution for export full csv-

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------

2. I have installed 8.0 branch as suggested by you after that I have build code and getting error in build (i.e "function not found ") and I analysis error code  with below mentioned steps:
1. As error file name showing in console I have analysis development code file in path (target/enhancedPlugin.js) and realised below mentioned webpack function convert in build like "s" or any character and getting error on production "s() function not found".

"_class EnhancedTablePlugin {
  constructor(initializerContext) {
    _babel_runtime_helpers_defineProperty__WEBPACK_IMPORTED_MODULE_0___default()(this, "initializerContext", void 0);
    _babel_runtime_helpers_defineProperty__WEBPACK_IMPORTED_MODULE_0___default()(this, "createBaseVisualization", void 0);
    this.initializerContext = initializerContext;
  }_ "

Solution : after that I have remove this function in particular file in build and run plugin successfully.

3. Sir, my custom plugin those are developed after customised "vis_type_markdown"  plugin by me and run successfully on development and build successfully. But getting error in production "function not found" and analysis my plugin used many webpack  babel runtime function in "target/myplugin.js" file instead or your enhanced plugin( enhanced plugin used only one one function as mentioned above).
And I have found in difference in both plugins and observed that enhanced plugin developed with angular + React then that are not used more than one webpack runtime helper function but my custom plugin is developed in only React and used many elastic and kibana library then that are used many webpack runtime helper functions example as mentioned below:


const MarkdownVisComponentCustom = /*#__PURE__*/Object(react__WEBPACK_IMPORTED_MODULE_1__["lazy"])(() => __webpack_require__.e(/*! import() */ 0).then(__webpack_require__.bind(null, /*! ./markdown_vis_custom_controller */ "./public/markdown_vis_custom_controller.tsx")));
const markdownVisRenderer = {
  name: 'markdown_vis_custom',
  displayName: 'markdown visualization',
  reuseDomNode: true,
  render: async (domNode, {
    visParams
  }, handlers) => {
    handlers.onDestroy(() => {
      Object(react_dom__WEBPACK_IMPORTED_MODULE_2__["unmountComponentAtNode"])(domNode);
    });
    Object(react_dom__WEBPACK_IMPORTED_MODULE_2__["render"])(Object(_emotion_react__WEBPACK_IMPORTED_MODULE_4__["jsx"])(_src_plugins_visualizations_public__WEBPACK_IMPORTED_MODULE_3__["VisualizationContainer"], {
      className: "markdownVis",
      handlers: handlers
    }, Object(_emotion_react__WEBPACK_IMPORTED_MODULE_4__["jsx"])(MarkdownVisComponentCustom, **_babel_runtime_helpers_extends__WEBPACK_IMPORTED_MODULE_0___default**()({}, visParams, {
      renderComplete: handlers.done
    }))), domNode);
  }
};

 used in this function  --- _**babel_runtime_helpers_extends__WEBPACK_IMPORTED_MODULE_0___default** -- in developement and getting error on production.

Please help me on above mentioned issues.

**Comment by fbaligand** [comment]
Hi,

1. I have no idea of what you call "secure". I can just tell you that it allows to write file on disk, using several chunks. Concretely, I make several XHR request to Kibana servers to get data by chunk of 10000 lines and then I write it to disk using StreamSaver. Finally, it works fine on Chrome (and so all webkit-based browsers), but not on Firefox.
2. Well, honestly, I never look after "compiled" js code and I don't think that is a very good idea. And so, I can't help you on generated code by "compile" task. That said, I precise that for my build for Kibana 8.1, I use enhanced-table 8.0 branch, but Kibana 8.1 branch.
